# 4.6.2 Combining and Exporting Data

## This script contains the following points:

#### 1. Import the data sets into Jupyter
#### 2. Check the dimensions of the imported dataframes
#### 3. Determine a suitable way to combine the orders_products_combined dataframe with the products data set
#### 4. Confirm the results of the merge using the merge flag
#### 5. Export the newly created dataframe as ords_prods_merge in a suitable format

In [24]:
# Import libraries
import pandas as pd
import numpy as np
import os

### 1. Import the data sets into Jupyter

In [25]:
path = r'/Users/test/Desktop/Data Analysis/11-2024 Instacart Basket Analysis'

In [26]:
# Import dataset orders_products_combined.pkl
df_ords_prods_combined = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_combined.pkl'))

In [27]:
# Import dataset products_rechecked.csv
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_rechecked.csv'), index_col = False)

### 2. Check the dimensions of the imported dataframes

In [29]:
# Check the output
df_ords_prods_combined.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                     NaN         True         196                  1   
1                     NaN         True       14084                  2   
2                     NaN         True       12427                  3   
3                     NaN         True       26088                  4   
4                     NaN         True       26405                  5   

   reordered _merge  
0          0   both  
1          0   both  
2          0   both  
3          0   both  
4          0   both

In [30]:
# Check the shape of the imported dataframe
df_ords_prods_combined.shape

(32434489, 11)

In [31]:
# Check the output
df_prods.head()

Unnamed: 0  product_id                                       product_name  \
0           0           1                         Chocolate Sandwich Cookies   
1           1           2                                   All-Seasons Salt   
2           2           3               Robust Golden Unsweetened Oolong Tea   
3           3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4           4           5                          Green Chile Anytime Sauce   

   aisle_id  department_id  prices  
0        61             19     5.8  
1       104             13     9.3  
2        94              7     4.5  
3        38              1    10.5  
4         5             13     4.3

In [32]:
# Remove unnecessary columns
df_prods = df_prods.drop(['Unnamed: 0'], axis=1)

In [33]:
# Check the output again
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [34]:
# Check the shape
df_prods.shape

(49670, 5)

### 3. Determine a suitable way to combine the orders_products_combined dataframe with the products data set

As both of the dataframes have the column 'product_id', it can be used as key column. A default inner merge would be a suitable way.

Since the previous merge operation created a 'merge' column as a flag, it will no longer be useful for the new merge between 'orders_products_combined' and 'df_prods'. For better presentation, the 'merge' column from the previous merge should be dropped after the new merge is completed.

In [35]:
# Drop 'merge' column
df_ords_prods_combined = df_ords_prods_combined.drop(['_merge'], axis=1)

In [36]:
# Check output
df_ords_prods_combined.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                     NaN         True         196                  1   
1                     NaN         True       14084                  2   
2                     NaN         True       12427                  3   
3                     NaN         True       26088                  4   
4                     NaN         True       26405                  5   

   reordered  
0          0  
1          0  
2          0  
3          0  
4          0

In [37]:
df_ords_prods_combined.shape

(32434489, 10)

In [38]:
# Merge the updated dataframes
df_ords_prods_merge = df_prods.merge(df_ords_prods_combined, on = 'product_id', indicator = True)

In [39]:
# Check the output
df_ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   days_since_prior_order  first_order  add_to_cart_order  reordered _merge  
0                     3.0        False                  5          0   both  
1                    20.0        False                  1          1   both  
2                     6.0        False                 20          0   both  
3                     NaN         True                 10          0   both  
4                     9.0        False                 11          1   both

In [40]:
df_ords_prods_merge.shape

(32399732, 15)

### 4. Confirm the results of the merge using the merge flag

In [41]:
df_ords_prods_merge['_merge'].value_counts()

_merge
both          32399732
left_only            0
right_only           0
Name: count, dtype: int64

The dataframe (after the merge) has 32,399,732 rows. Each of those rows have values found in both dataframes, as we used an inner join for the purposes of this project.

### 5. Export the newly created dataframe as ords_prods_merge in a suitable format

In [42]:
# Export data to pkl
df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_merge.pkl'))